apps/sync/js/adapters/history.js

```diff
@@ -168,12 +189,66 @@ var HistoryHelper = (() => {
     });
   }
 
+  function checkIfCleared(userid) {
+    return getLastRevisionId(userid).then(lastRevisionId => {
+      return _ensureStore().then(store => {
+        if (lastRevisionId === null) {
+          // Skip first task which is always { id: null, operation: 'clear' }
+          var cursor = store.sync();
+          cursor.next().then(() => {
+            return cursor;
+          });
+        }
+        return store.sync(lastRevisionId);
+      });
+    }).then(cursor => {
+      var hasCleared = false;
+      return new Promise(resolve => {
+        function runNextTask(cursor) {
+          cursor.next().then(task => {
+            if (task.operation === 'done') {
+              resolve({
+                revisionId: task.revisionId,
+                hasCleared
+              });
+            } else {
+              if (task.operation === 'clear') {
+                hasCleared = true;
+              }
+              // Avoid stack overflow:
+              setTimeout(() => {
+                // Will eventually get to a 'done' task:
+                runNextTask(cursor);
+              });
+            }
+          });
+        }
+        runNextTask(cursor);
+      });
+    });
+  }
+
+  function handleClear(userid) {
+    var lastRevisionId;
+    return checkIfCleared(userid)
+    .then(result => {
+      lastRevisionId = result.revisionId;
+      if(result.hasCleared) {
+        return removeSyncedCollectionMtime(userid);
```

Comment:
It would be good to have a comment explaining the overall clearing process. The adapters are becoming quite complex for newcomers.